features/preferences/impl/src/main/kotlin/io/element/android/features/preferences/impl/notifications/NotificationSettingsPresenter.kt

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2023 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.features.preferences.impl.notifications
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.collectAsState
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
+import io.element.android.libraries.architecture.Presenter
+import io.element.android.libraries.matrix.api.MatrixClient
+import io.element.android.libraries.matrix.api.notificationsettings.NotificationSettingsService
+import io.element.android.libraries.matrix.api.room.RoomNotificationMode
+import io.element.android.libraries.pushstore.api.UserPushStore
+import io.element.android.libraries.pushstore.api.UserPushStoreFactory
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+import kotlin.time.Duration.Companion.seconds
+
+class NotificationSettingsPresenter @Inject constructor(
+    private val notificationSettingsService: NotificationSettingsService,
+    private val userPushStoreFactory: UserPushStoreFactory,
+    private val matrixClient: MatrixClient,
+    private val systemNotificationsEnabledProvider: SystemNotificationsEnabledProvider
+) : Presenter<NotificationSettingsState> {
+    @Composable
+    override fun present(): NotificationSettingsState {
+        val userPushStore = userPushStoreFactory.create(matrixClient.sessionId)
+        val systemNotificationsEnabled: MutableState<Boolean> = remember {
+            mutableStateOf(systemNotificationsEnabledProvider.notificationsEnabled())
+        }
+
+        val localCoroutineScope = rememberCoroutineScope()
+        val appNotificationsEnabled = userPushStore
+            .getNotificationEnabledForDevice()
+            .collectAsState(initial = false)
+
+        val matrixSettings: MutableState<NotificationSettingsState.MatrixSettings> = remember {
+            mutableStateOf(NotificationSettingsState.MatrixSettings.Uninitialized)
+        }
+
+        LaunchedEffect(Unit) {
+            fetchSettings(matrixSettings)
```

Comment:
Maybe this first `fetchSettings` could be moved inside `observeNotificationSettings` and returned as the first value? I'm not sure how feasible it is with the `debounce` there though.

Reply:
Kind of like Rx's [startWidth](https://reactivex.io/documentation/operators/startwith.html). Not quite sure how to do with flow or if it would make much easier to understand.

Reply:
There is also a `startWith` function for Flows, but it's deprecated. Ideally, it should be something like `.onStart { emit(initialValue) }`, but it might not be worth the trouble.